File: codeforces/1800~1899/1860/c.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<int> p(n);
  for (int i = 0; i < n; ++i) cin >> p[i];

  int ans = 0;
  set<int> lose, win;
  for (int i = 0; i < n; ++i) {
    auto it = win.upper_bound(p[i]);
    if (it != win.begin()) {
      lose.insert(p[i]);
    } else {
      it = lose.upper_bound(p[i]);
      if (it != lose.begin()) {
        win.insert(p[i]);
        ans++;
      } else {
        lose.insert(p[i]);
      }
    }
  }

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1800~1899/1860/c.rs
use std::io;
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BTreeSet;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let n = read_int::<i32>();
    let p = read_vec::<i32>();
    let mut lose: BTreeSet<i32> = BTreeSet::new();
    let mut win: BTreeSet<i32> = BTreeSet::new();

    let mut ans = 0;
    for i in 0..n {
        let x = p[i as usize];
        if !win.is_empty() && *win.first().unwrap() < x {
            lose.insert(x.clone());
        } else {
            if !lose.is_empty() && *lose.first().unwrap() < x {
                win.insert(x.clone());
                ans += 1;
            } else {
                lose.insert(x.clone());
            }
        }
    }

    println!("{}", ans);
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
